src/platform/packages/shared/kbn-zod-helpers/src/deep_strict.ts

```diff
@@ -7,321 +7,54 @@
-
-    if (isDangerousParsableSchema(schema)) {
-      dangerous.add({ key, schema });
-      return;
-    }
-
-    const def = schema._def;
-
-    switch (def.typeName) {
-      case ZodFirstPartyTypeKind.ZodObject:
-        return Object.entries(def.shape()).forEach(([prop, innerSchema]) =>
-          innerAssertAllParsableSchemas(innerSchema as z.Schema, key ? `${key}.${prop}` : prop)
-        );
-      case ZodFirstPartyTypeKind.ZodOptional:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-      case ZodFirstPartyTypeKind.ZodArray:
-        return innerAssertAllParsableSchemas(def.type, key);
-
-      case ZodFirstPartyTypeKind.ZodUnion:
-      case ZodFirstPartyTypeKind.ZodDiscriminatedUnion:
-        const schemas: z.Schema[] = def.options;
-        return schemas.forEach((innerSchema) => innerAssertAllParsableSchemas(innerSchema, key));
-
-      case z.ZodFirstPartyTypeKind.ZodIntersection:
-        innerAssertAllParsableSchemas(def.left, key);
-        innerAssertAllParsableSchemas(def.right, key);
-        return;
-
-      case z.ZodFirstPartyTypeKind.ZodEffects:
-        if (def.effect.type === 'transform') {
-          dangerous.add({ key, schema: schema as ZodEffects<ZodAny> });
-        }
-        return innerAssertAllParsableSchemas(def.schema, key);
-
-      case z.ZodFirstPartyTypeKind.ZodRecord:
-        return innerAssertAllParsableSchemas(def.valueType, key);
-
-      case ZodFirstPartyTypeKind.ZodDefault:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-
-      case ZodFirstPartyTypeKind.ZodLazy:
-        return innerAssertAllParsableSchemas(def.getter(), key);
-
-      case ZodFirstPartyTypeKind.ZodNullable:
-        return innerAssertAllParsableSchemas(def.innerType, key);
-    }
+import { z, ZodIssueCode } from '@kbn/zod';
+import { difference, isArray, isPlainObject } from 'lodash';
+
+function getFlattenedKeys(obj: unknown, parentKey = '', keys: Set<string> = new Set()) {
+  if (isPlainObject(obj)) {
+    Object.entries(obj as Record<string, any>).forEach(([key, value]) => {
+      getFlattenedKeys(value, parentKey ? `${parentKey}.${key}` : key, keys);
+    });
+  } else if (isArray(obj)) {
+    obj.forEach((value) => {
+      getFlattenedKeys(value, parentKey, keys);
+    });
+  } else {
+    keys.add(parentKey);
   }
+  return keys;
 }
 
-function getHandlingSchemas(schema: z.Schema, key: string, value: object): z.Schema[] {
-  assertIsParsableSchema(schema, key);
-
-  if (isPrimitiveParsableSchema(schema) || isDangerousParsableSchema(schema)) {
-    return [];
+function parseStrict<TSchema extends z.Schema>(
+  source: z.Schema,
+  input: z.ParseInput
+): z.ParseReturnType<z.output<TSchema>> {
+  const next = source._parse(input);
+  if (!z.isValid(next)) {
+    return next;
   }
 
-  const def = schema._def;
+  const allInputKeys = Array.from(getFlattenedKeys(input.data).values());
+  const allOutputKeys = Array.from(getFlattenedKeys(next.value as Record<string, any>).values());
```

Comment:
nit: A `Set` might be directly passed to `Array.from()` meaning `.values()` isn't necessary here.

Reply:
fair!